File: solutions/ruby/zebra-puzzle/1/zebra_puzzle.rb
class ZebraPuzzle
  private_class_method :new

  def self.water_drinker
    RESULT[0]
  end

  def self.zebra_owner
    RESULT[1]
  end

  private

  def solve_puzzle
    # Las variables serán índices 0-4 (posición de la casa)
    [0, 1, 2, 3, 4].permutation do |eng, spa, ukr, nor, jap|
      next unless nor == 0                    # 10. Norwegian en casa 0

      [0, 1, 2, 3, 4].permutation do |red, green, ivory, yellow, blue|
        next unless eng == red                # 2. Englishman en casa roja
        next unless green == ivory + 1        # 6. Green a la derecha de Ivory
        next unless (nor - blue).abs == 1   # 15. Norwegian junto a Blue

        [0, 1, 2, 3, 4].permutation do |coffee, tea, milk, oj, water|
          next unless milk == 2               # 9. Leche en el centro (casa 2)
          next unless green == coffee         # 4. Green casa bebe Coffee
          next unless ukr == tea              # 5. Ukrainian bebe Tea

          [0, 1, 2, 3, 4].permutation do |dancing, painting, reading, football, chess|
            next unless yellow == painting    # 8. Yellow casa es Painter
            next unless oj == football        # 13. OJ bebe Football
            next unless jap == chess          # 14. Japanese juega Chess

            [0, 1, 2, 3, 4].permutation do |dog, snail, fox, horse, zebra|
              next unless spa == dog          # 3. Spaniard tiene Dog
              next unless snail == dancing    # 7. Snail owner baila
              next unless (reading - fox).abs == 1   # 11. Reading junto a Fox
              next unless (painting - horse).abs == 1 # 12. Painting junto a Horse

              # Mapear posición -> nacionalidad
              pos_to_nation = {
                eng => 'Englishman',
                spa => 'Spaniard',
                ukr => 'Ukrainian',
                nor => 'Norwegian',
                jap => 'Japanese'
              }

              return [pos_to_nation[water], pos_to_nation[zebra]]
            end
          end
        end
      end
    end
  end

  RESULT = new.send(:solve_puzzle).freeze
end

File: solutions/javascript/zebra-puzzle/1/zebra-puzzle.js
// A helper function to generate all 120 permutations of a 5-element array
const getPermutations = (arr) => {
  if (arr.length === 0) return [[]];
  const result = [];
  for (let i = 0; i < arr.length; i++) {
    const rest = getPermutations([...arr.slice(0, i), ...arr.slice(i + 1)]);
    for (let j = 0; j < rest.length; j++) {
      result.push([arr[i], ...rest[j]]);
    }
  }
  return result;
};

// Generate our base permutations for 5 houses (Indices 0 to 4)
const PERMUTATIONS = getPermutations([0, 1, 2, 3, 4]);

export class ZebraPuzzle {
  constructor() {
    this._waterDrinker = '';
    this._zebraOwner = '';
    this.solve();
  }

  solve() {
    const nationalitiesList = ['Englishman', 'Spaniard', 'Ukrainian', 'Norwegian', 'Japanese'];

    // We map arrays to represent house indices (0-4) for each property.
    // E.g., `c[4]` represents the house index of the Blue house.
    
    for (const c of PERMUTATIONS) {
      // Colors: 0:Red, 1:Green, 2:Ivory, 3:Yellow, 4:Blue
      if (c[4] !== 1) continue; // Norwegian is 0, is next to Blue. So Blue = 1
      if (c[1] !== c[2] + 1) continue; // Green is immediately right of Ivory

      for (const n of PERMUTATIONS) {
        // Nationalities: 0:Englishman, 1:Spaniard, 2:Ukrainian, 3:Norwegian, 4:Japanese
        if (n[3] !== 0) continue; // Norwegian lives in first house
        if (n[0] !== c[0]) continue; // Englishman lives in Red

        for (const d of PERMUTATIONS) {
          // Drinks: 0:Coffee, 1:Tea, 2:Milk, 3:OrangeJuice, 4:Water
          if (d[2] !== 2) continue; // Middle house (index 2) drinks Milk
          if (d[0] !== c[1]) continue; // Green house drinks Coffee
          if (d[1] !== n[2]) continue; // Ukrainian drinks Tea

          for (const h of PERMUTATIONS) {
            // Hobbies: 0:Dancing, 1:Painter, 2:Reading, 3:Football, 4:Chess
            if (h[1] !== c[3]) continue; // Painter is in Yellow house
            if (h[4] !== n[4]) continue; // Japanese plays Chess
            if (h[3] !== d[3]) continue; // Football player drinks OJ

            for (const p of PERMUTATIONS) {
              // Pets: 0:Dog, 1:Snails, 2:Fox, 3:Horse, 4:Zebra
              if (p[0] !== n[1]) continue; // Spaniard owns Dog
              if (p[1] !== h[0]) continue; // Snails owner dances
              if (Math.abs(h[2] - p[2]) !== 1) continue; // Reading next to Fox
              if (Math.abs(h[1] - p[3]) !== 1) continue; // Painter next to Horse

              // If we make it here, all constraints are satisfied!
              const waterHouseIndex = d[4];
              const zebraHouseIndex = p[4];

              // Find which nationality lives at those specific house indices
              this._waterDrinker = nationalitiesList[n.indexOf(waterHouseIndex)];
              this._zebraOwner = nationalitiesList[n.indexOf(zebraHouseIndex)];
              
              return; // We found the unique solution, no need to keep searching
            }
          }
        }
      }
    }
  }

  waterDrinker() {
    return this._waterDrinker;
  }

  zebraOwner() {
    return this._zebraOwner;
  }
}
